experiments/evaluation.py

```diff
@@ -0,0 +1,156 @@
+import json
+from typing import List, Tuple
+
+import numpy as np
+import pandas as pd
+
+from haystack.schema import Answer
+from haystack.nodes.evaluator.evaluator import semantic_answer_similarity
+from rest_api.schema import QuestionAnswerPair, PipelineHyperParams
+from rest_api.utils import get_pipelines
+from haystack import Pipeline
+from document_indexing.s3_storage import S3Storage
+from experiments.wandb_logger import WandBLogger
+
+
+class PipelineEvaluation:
+    @staticmethod
+    def ranked_answers(answers: List[Answer]) -> List[str]:
+        return [a.answer for a in sorted(answers, key=lambda a: a.score) if a.context]
+
+    @staticmethod
+    def get_n_worst_best_examples(
+        questions: List[str], scores: List[float], retrieved_questions: List[List[str]], n: int = 3
+    ) -> Tuple[pd.DataFrame, pd.DataFrame]:
+        sorted_scores_index = np.argsort(scores)
+        index_worst = sorted_scores_index[:n]
+        index_best = sorted_scores_index[-n:]
+        k = len(retrieved_questions[0])
+        worst = pd.DataFrame(
+            data=[retrieved_questions[i] for i in index_worst], columns=[f"rank{i}" for i in range(1, k + 1)]
+        )
+        worst["question"] = [questions[i] for i in index_worst]
+        best = pd.DataFrame(
+            data=[retrieved_questions[i] for i in index_best], columns=[f"rank{i}" for i in range(1, k + 1)]
+        )
+        best["question"] = [questions[i] for i in index_best]
+        return worst, best
+
+    def get_performance_metrics(
+        self, question_answer_pairs: List[QuestionAnswerPair], pipeline_hyper_params: PipelineHyperParams
+    ):
+        query_pipeline: Pipeline = get_pipelines(pipeline_hyper_params).get("query_pipeline", None)
+        true_questions = []
+        true_answers = []
+        faq_answers = []
+        extractive_answers = []
+
+        # for each question answering pair, query the pipeline
+        for q_and_a_pair in question_answer_pairs:
+            true_answers.append([q_and_a_pair.answer])
+            true_questions.append(q_and_a_pair.question)
+
+            # faq
+            result_faq = query_pipeline.run(query=q_and_a_pair.question, params={"CustomClassifier": {"index": "faq"}})
```

Comment:
Shouldn't it be here: `q_and_a_pair.alternative_question`, instead of `q_and_a_pair.question`? During evaluation we'd like to query the faq model with alternative versions of the questions, and see if the model correctly retrieves original question (and answer to it ofc)

Reply:
yes can do!

Reply:
https://github.com/rafaljanwojcik/board-game-rules-explainer/pull/4/commits/4000f1df69511fabeefbf403d4cba1dcfdea53ce